### PR TITLE
Make byte blocks more even in read_bytes

### PR DESCRIPTION
**Pull request by martindurant**
This avoids the chance of having a final block with just a few
useless bytes in, by shrinking the blocksize to make all the
blocks roughly equal

- [x] ~Closes #xxxx~
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

cc @douglasdavis , this would solve one half of the "empty array" issue.

### CODE REVIEW

**Comment by martindurant** [comment]
Note that I commented a couple of lines of test rather than removed, just so it was obvious what I was doing. I can remove before merger.

**Comment by martindurant** [comment]
ping @dask/maintenance , any thoughts?

File: dask/bytes/core.py [hunk 1]
```diff
@@ -112,13 +112,31 @@ def read_bytes(
...
                 )
-            off = list(range(0, size, blocksize))
-            length = [blocksize] * len(off)
-            if not_zero:
-                off[0] = 1
-                length[0] -= 1
-            offsets.append(off)
-            lengths.append(length)
+
+            if size:
```
Comment (GenevieveBuckley) [inline dask/bytes/core.py:116]: Should this if statement be joined to the clause above? 
```python
if size is None:
    raise ValueError(...
else:
   # shrink blocksize to give same number of parts
    ...
```

File: dask/bytes/core.py [hunk 2]
```diff
@@ -112,13 +112,31 @@ def read_bytes(
...
+                # shrink blocksize to give same number of parts
+                if size % blocksize and size > blocksize:
+                    blocksize = size / (size // blocksize)
```
Comment (GenevieveBuckley) [inline dask/bytes/core.py:119]: This seems to me like a nice way to handle it

File: dask/bytes/core.py [hunk 3]
```diff
@@ -112,13 +112,31 @@ def read_bytes(
...
+                    off.append(int(place))
+                    length.append(off[-1] - off[-2])
+                length.append(size - off[-1])
+
+                if not_zero:
+                    off[0] = 1
+                    length[0] -= 1
+                offsets.append(off)
+                lengths.append(length)
+            else:
```
Comment (GenevieveBuckley) [inline dask/bytes/core.py:136]: I think this clause only gets activated when `size=0`, is that right? At first I was a bit confused about how this clause would get triggered, given the `if size is None` check above. If the number zero is the only other case which would return False, then I think we need to put all three conditions into the same if then else statement.

Comment (GenevieveBuckley) [reply dask/bytes/core.py:136]:
```python
if size is None:
    raise ValueError(
        ...
elif size == 0:
    # skip empty
    ...
else:
   # shrink blocksize to give same number of parts
    ...
```

**Review by GenevieveBuckley — APPROVED** [review]
I like this solution. It seems to work pretty well with the tiny toy examples I tried it out on.

Left a couple of minor comments about improving the clarity of the code, nothing major.

**Comment by GenevieveBuckley** [comment]
I notice the GPU CI is failing, I'm not entirely sure why. It looks like more a problem with the CI than your code? The NVIDIA crew are all pretty busy this time of year, might be worth checking to see if something similar is happening on other recent PRs.

**Comment by martindurant** [comment]
> It looks like more a problem with the CI than your code

As far as I can tell. - and not sure I have full access to the logs here - the tests simply didn't run. Which is fine, this doesn't touch anything GPU-specific, just chopping up text files.

File: dask/bytes/core.py (hunk 3 above)
Comment (martindurant) [reply dask/bytes/core.py:136]: Yes, your understanding is correct, I'll move blocks around for clarity.

File: dask/bytes/core.py (hunk 1 above)
Comment (martindurant) [reply dask/bytes/core.py:116]: Covered below

**Comment by GenevieveBuckley** [comment]
That's what it looks like to me as well. I am +1 on ignoring whatever this weirdness is, it doesn't seem to have anything to do with the actual PR.

**Comment by martindurant** [comment]
Windows test is failing on numpy import in workers?!

```
  File "C:\Miniconda3\envs\test-environment\lib\site-packages\numpy\core\overrides.py", line 7, in <module>

    from numpy.core._multiarray_umath import (

ImportError: DLL load failed: The paging file is too small for this operation to complete.
```

File: dask/bytes/core.py [hunk 4]
```diff
@@ -112,13 +112,31 @@ def read_bytes(
...
-            lengths.append(length)
+
+            elif size == 0:
+                # skip empty
+                offsets.append([])
+                lengths.append([])
+            else:
+                # shrink blocksize to give same number of parts
+                if size % blocksize and size > blocksize:
+                    blocksize = size / (size // blocksize)
```
Comment (martindurant) [inline dask/bytes/core.py:123]: Please check: I think if I rewrite `blocksize` here, I risk reducing it on every file iteration, right? I should have a new local variable here, right?

Comment (GenevieveBuckley) [reply dask/bytes/core.py:123]: Oh, yes! Sorry I didn't catch that, you're right.

**Pull request merged**